#! /bin/sh /usr/share/dpatch/dpatch-run
## 053_mod_cache_HEAD_41230.dpatch
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: fix http://issues.apache.org/bugzilla/show_bug.cgi?id=41230

@DPATCH@
--- httpd/modules/cache/mod_cache.c.orig	2007-04-11 13:29:14.000000000 +0200
+++ httpd/modules/cache/mod_cache.c	2007-04-11 14:06:29.000000000 +0200
@@ -456,7 +456,7 @@ static int cache_save_filter(ap_filter_t
          */
         reason = "No Last-Modified, Etag, or Expires headers";
     }
-    else if (r->header_only) {
+    else if (r->header_only && !cache->stale_handle) {
         /* HEAD requests */
         reason = "HTTP HEAD request";
     }
@@ -589,11 +589,12 @@ static int cache_save_filter(ap_filter_t
             cache->provider->remove_entity(cache->stale_handle);
             /* Treat the request as if it wasn't conditional. */
             cache->stale_handle = NULL;
+            rv = !OK;
         }
     }
 
-    /* no cache handle, create a new entity */
-    if (!cache->handle) {
+    /* no cache handle, create a new entity only for non-HEAD request */
+    if (!cache->handle && !r->header_only) {
         rv = cache_create_entity(r, size);
         info = apr_pcalloc(r->pool, sizeof(cache_info));
         /* We only set info->status upon the initial creation. */
